### ISSUE TITLE
Misleading behavior on matrix and vector division, which is in contradiction with the documentation in nn.softmax

### ISSUE THREAD

**Issue by T-Almeida** [tensorflow/tensorflow#53112]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- TensorFlow version (use command below): 2.7.0
- Python version: 3.7.12

**Describe the current behavior**

I found a misleading behavior when using broadcast between matrix and vectors. I will use the example of softmax to demonstrate the issue, which if it's not a bug then the following documentation I believe it is wrong " https://www.tensorflow.org/api_docs/python/tf/nn/softmax".

```
# random matrix
A = tf.random.uniform((4,4), maxval=10, dtype=tf.float64)

# softmax by hand, as suggested in: https://www.tensorflow.org/api_docs/python/tf/nn/softmax
A_softmax = tf.exp(A) / tf.reduce_sum(tf.exp(A), axis=-1)
tf.reduce_sum(A_norm, axis=-1) 
# >>> <tf.Tensor: shape=(4,), dtype=float64, numpy=array([0.79724022, 0.34883704, 3.8507741 , 2.37962821])>

A_nn_softmax = tf.nn.softmax(A, axis=-1)
tf.reduce_sum(A_nn_softmax, axis=-1) 
# >>> <tf.Tensor: shape=(4,), dtype=float64, numpy=array([1., 1., 1., 1.])>
```
In the above example, I use the tf.nn.softmax as a baseline and then use the corresponding computation as described in the documentation ( that is also the way I usually implement the softmax operation ). However, as you can see, A_softmax (direct computation) and A_nn_softmax (from the function nn.softmax) gave two different outputs, which according to the documentation should have not happened. I expect the same output for both matrices.

After some time I realise that the problem was related to the column-wise division that has, what I believe to be, a **misleading** behaviour.

```
tf.exp(A) / tf.reduce_sum(tf.exp(A), axis=-1)
```

In the above computation, I am dividing a (4, 4) matrix per (4,) vector, which I expect to perform a column-wise division since it is a column vector. However, it performs a row-wise division. I would like to know if this is an intended behavior and if so, I believe that the documentation of nn.softmax is wrong and should be "tf.exp(logits) / tf.reduce_sum(tf.expand_dims(tf.exp(logits), axis=-1), axis=-1)".

**Describe the expected behavior**

I expect that A_softmax equals A_nn_softmax.

To achieve the expected behavior I need to implicitly add an additional dimension to the denominator vector, becoming (4, 1).

In code correspond to:

```
tf.exp(logits) /tf.expand_dims(tf.reduce_sum(tf.exp(logits), axis=-1), axis=-1)
```
In alternative I can also use "keepdims=True" when doing the reduce_sum

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1Iv7cw4_gddfaxre5AD_oBQlGih6cJvMM?usp=sharing

**Comment by T-Almeida** [comment]
I was able to run more experiments, and now I believe that is indeed the intended behavior and after all, what was misleading was the documentation in nn.softmax "tf.exp(logits) / tf.reduce_sum(tf.exp(logits), axis)", since is wrong when axis=-1 which is the default behavior.

Maybe change this to a documentation issue?

**Comment by tilakrayal** [comment]
@T-Almeida ,
I ran the code shared and face a different error, please find the gist [here](https://colab.research.google.com/gist/tilakrayal/f4bf945d7416ff13e6c2f2ffda8f81e5/2-7.ipynb) and share all dependencies to replicate the issue.

**Comment by T-Almeida** [comment]
Hi, indeed I change the code and forgot to update that line, my mistake. 

Instead of "tf.reduce_sum(A_norm, axis=-1)" it should be "tf.reduce_sum(A_softmax, axis=-1)". Then the output of this should be ones (correct normalization sums to 1), as suggested in https://www.tensorflow.org/api_docs/python/tf/nn/softmax . However, you will see that this is not true because the vector resulting from the reduce_sum operation is handled as a row-vector (I believe). Therefore, I consider this to be an issue related to the documentation. Nevertheless, I found that to be also misleading since a column-wise reduce operation results in a row-vector.

**Comment by sanatmpa1** [comment]
@T-Almeida,

In this case of documentation bug, Would you like to submit a PR addressing the issue? Thanks!

**Comment by T-Almeida** [comment]
Hi, @sanatmpa1 yes sure, I can open a PR.

**Comment by sanatmpa1** [comment]
@T-Almeida,

As the PR that you have submitted is merged, Can you confirm if we are good to close this issue? Thanks!
